keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/ArtifactUsedConstraintEvaluator.kt

```diff
@@ -0,0 +1,32 @@
+package com.netflix.spinnaker.keel.constraints
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.core.api.ArtifactUsedConstraint
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+@Component
+class ArtifactUsedConstraintEvaluator(
+  override val eventPublisher: ApplicationEventPublisher
+) : ConstraintEvaluator<ArtifactUsedConstraint> {
+  override val supportedType = SupportedConstraintType<ArtifactUsedConstraint>("artifact-used")
+  override fun isImplicit() = true
+
+  override fun canPromote(artifact: DeliveryArtifact, version: String, deliveryConfig: DeliveryConfig, targetEnvironment: Environment): Boolean {
+    val allowedArtifacts = mutableSetOf<DeliveryArtifact>()
+    deliveryConfig
+      .environments
+      .firstOrNull { it.name == targetEnvironment.name }
+      ?.resources
+      ?.forEach { resource ->
+        resource.findAssociatedArtifact(deliveryConfig)
+          ?.let { usedArtifact ->
+            allowedArtifacts.add(usedArtifact)
+          }
+      }
+
+    return artifact in allowedArtifacts
```

Comment:
If you wanted to increase your functional programming score, I think you can rewrite this function like this:

```kotlin
  override fun canPromote(artifact: DeliveryArtifact, version: String, deliveryConfig: DeliveryConfig, targetEnvironment: Environment): Boolean {
    return deliveryConfig
      .environments
      .firstOrNull { it.name == targetEnvironment.name }
      ?.resources
      ?.map { it.findAssociatedArtifact(deliveryConfig) }
      ?.any { it == artifact }
      ?: false
  }
```